lisa/sut_orchestrator/hyperv/platform_.py

```diff
@@ -0,0 +1,317 @@
+            assert self._source_vhd
+
+            node.name = vm_name
+
+            node_context = get_node_context(node)
+            node_context.vm_name = vm_name
+            node_context.host = self._server
+
+            node_context.working_path = PureWindowsPath(
+                self._server.working_path / f"{vm_name}"
+            )
+
+            vm_vhd_name = f"{vm_name}{self._source_vhd.suffix}"
+            vhd_path = PureWindowsPath(node_context.working_path / f"{vm_vhd_name}")
+
+            self._server.tools[Mkdir].create_directory(str(node_context.working_path))
+
+            self._server.tools[Cp].copy(self._source_vhd, vhd_path)
+
+            self._resize_vhd_if_needed(vhd_path, node_runbook)
+
+            assert isinstance(node.capability.core_count, int)
+            assert isinstance(node.capability.memory_mb, int)
+
+            com1_pipe_name = f"{vm_name}-com1"
+            com1_pipe_path = f"\\\\.\\pipe\\{com1_pipe_name}"
+
+            log.info(f"Serial logs at {node_context.console_log_path}")
+            node_context.serial_log_task_mgr = self._console_logger.start_logging(
+                com1_pipe_name, node_context.console_log_path, log
+            )
+
+            hv.create_vm(
+                name=vm_name,
+                guest_image_path=str(vhd_path),
+                switch_name=default_switch.name,
+                generation=node_runbook.hyperv_generation,
+                cores=node.capability.core_count,
+                memory=node.capability.memory_mb,
+                secure_boot=False,
+                com_ports={
+                    1: com1_pipe_path,
+                },
+                extra_args=extra_args,
+            )
+
+            ip_addr = hv.get_ip_address(vm_name)
+            username = self.runbook.admin_username
+            password = self.runbook.admin_password
+            node.set_connection_info(
+                address=ip_addr, username=username, password=password
+            )
+
+    def _resize_vhd_if_needed(
+        self, vhd_path: PureWindowsPath, node_runbook: HypervNodeSchema
+    ) -> None:
+        pwsh = self._server.tools[PowerShell]
+        vhd_size = int(pwsh.run_cmdlet(f"(Get-VHD -Path {vhd_path}).Size"))
+        if vhd_size < node_runbook.osdisk_size_in_gb * 1024 * 1024 * 1024:
+            pwsh.run_cmdlet(
+                f"Resize-VHD -Path {vhd_path} "
+                f"-SizeBytes {node_runbook.osdisk_size_in_gb * 1024 * 1024 * 1024}"
+            )
+
+    def _delete_environment(self, environment: Environment, log: Logger) -> None:
+        self._delete_nodes(environment, log)
+
+    def _delete_nodes(self, environment: Environment, log: Logger) -> None:
+        hv = self._server.tools[HyperV]
+        for node in environment.nodes.list():
+            node_ctx = get_node_context(node)
+            vm_name = node_ctx.vm_name
+
+            log.debug(f"Deleting VM {vm_name}")
+            hv.delete_vm(vm_name)
```

Comment:
Make sure it's async deletion by default. It would be fast.

Reply:
You mean don't wait for it to finish? Or do it in parallel?

Reply:
Both. Like Azure, use a flag to control whether wait for deletion.

Reply:
Looks like we don't have a `run_cmdlet_async()` in PowerShell tool, so that needs to be implemented first.

Reply:
https://github.com/microsoft/lisa/pull/3172

Reply:
Done.